src/vmm/src/config/builder.rs

```diff
@@ -0,0 +1,110 @@
+//! Config builder
+use std::convert::TryFrom;
+
+use super::{
+    BlockConfig, ConfigResult, ConversionError, KernelConfig, MemoryConfig, NetConfig, VMMConfig,
+    VcpuConfig,
+};
+
+/// Builder structure for VMMConfig
+#[derive(Debug)]
+pub struct Builder {
+    inner: ConfigResult<VMMConfig>,
+}
+
+impl Default for Builder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Builder {
+    pub fn new() -> Self {
+        Builder {
+            inner: Ok(VMMConfig::default()),
+        }
+    }
+
+    pub fn build(&self) -> ConfigResult<VMMConfig> {
```

Comment:
I've given it a bit more thought, and I don't think that there is really a nice way of disallowing defaults for `KernelConfig`. But, there is a hacky way :laughing: 

How about: we return an error here on `build`, when the path of the `KernelConfig` is empty? This would allow us to remove the `Default` implementation for `KernelConfig`, as well as returning an error in the builder as well when the required value is not set. Since this is not necessarily the nicest way of doing it, it would require documentation to explain why this hack is needed.

Reply:
When you say  remove the `Default` implementation, Do you also mean to say remove `[derive(Default)]` as well. That would make things complicated, because then we cannot use `derive(Default)` on the top level `VMMConfig`. 

I like the idea of throwing an `error` upon 'empty' kernel-config from the `build` function, but removing `Default` and/or `derive(Default)` would mean we cannot use `Default` for the entire `VMMConfig`, which would make everything mandatory (a bit of a UX annoyance) 

I believe this `build` should be a catchall place for all configuration errors by asserting values here, because this will indeed be the last in the chain (and a required one). Also, in the context of `clap` based CLI, this is not a huge problem, because missing kernel config will be reported as an error already, even before it comes here.

Reply:
I believe it's not possible to get rid of `Default` implementation for `KernelConfig` and hence at-least for now we are going ahead with catching errors in the `build` function as implemented. Hence marking this conversation as resolved.